### DISCUSSION TITLE
Criteria APIを使用した3列以上のin句作成

### DISCUSSION THREAD

**Discussion by ironclad-15256304** [domaframework/doma#968] [Q&A] [answered]
### 相談内容

いつもお世話になっております。
私の参加しているプロジェクトでは、下記のようなSQLを必要とするシチュエーションがしばしば存在します。

```
select t_0.*
  from table_a t_0
 where (t_0.key_1, t_0.key_2, t_0.key_3)
       in (
         (1,1,1), -- 条件1
         (1,1,2), -- 条件2
         (2,1,3), -- 条件3
       )
```

上記のように、3つ以上の列を条件としたin句を、可能であればCriteria APIのEntityqlを用いて作成したいのです。
※このとき、条件として使用する値はフロントエンドから渡ってくるようになっており、クエリ実行時点ではどこのテーブルにも存在していません。また、この後、joinとassociateを使用して関連するエンティティを取得します

DomaのCriteria API入門記事やサンプル、JavaDocを参照してみたのですが、1列に対するin句や、Tuple2を使用した2列に対するin句しか見つけることが出来ませんでした。
Criteria APIでは、このような形式のSQL作成はサポートされていないのでしょうか。

Criteria APIでの作成が不可能な場合、

1. 一旦トークンのようなキー項目と条件だけを格納したワークテーブルを作成し、それに対してexists句等で合致するデータを取得するようにする
2. SQLテンプレートを作成して実行する

のどちらかで対応したいと考えておりますが、何分頻繁に登場するため、可能であればCriteria APIを使用した実装を行いたいです。
お忙しいところ大変恐縮ですが、このようなシチュエーションでのベストプラクティスをご教示頂けますと幸いです。

### 動作環境

+ Java Framework: Spring boot 3.1.0
+ Doma: 2.53.1（doma-spring-boot 1.7.0）
+ DB: postres もしくは oracle

**Comment by nakamura-to** [comment, marked as answer]
残念ながら3列に対するIN句はサポートしていません。
代替手段としては、ANDやORを組み合わせて表現が可能です。

```java
Employee_ e = new Employee_();

Tuple3<Integer, Integer, String> t1 = new Tuple3<>(1, 10, "a");
Tuple3<Integer, Integer, String> t2 = new Tuple3<>(2, 20, "b");
List<Tuple3<Integer, Integer, String>> list = Arrays.asList(t1, t2);

List<Employee> result =
        entityql
                .from(e)
                .where(c -> {
                    for (Tuple3<Integer, Integer, String> t : list) {
                        c.and(() -> {
                            c.or(() -> c.eq(e.employeeId, t.getItem1()));
                            c.or(() -> c.eq(e.employeeNo, t.getItem2()));
                            c.or(() -> c.eq(e.employeeName, t.getItem3()));
                        });
                    }
                })
                .fetch();
```

Criteira APIを使った上記のクエリは次のようなSQLに変換されます。

```sql
select 
  t0_.EMPLOYEE_ID, t0_.EMPLOYEE_NO, t0_.EMPLOYEE_NAME, t0_.MANAGER_ID, t0_.HIREDATE, t0_.SALARY, t0_.DEPARTMENT_ID, t0_.ADDRESS_ID, t0_.VERSION 
from 
  EMPLOYEE t0_ 
where 
  ((t0_.EMPLOYEE_ID = 1) or (t0_.EMPLOYEE_NO = 10) or (t0_.EMPLOYEE_NAME = 'a')) 
  and 
  ((t0_.EMPLOYEE_ID = 2) or (t0_.EMPLOYEE_NO = 20) or (t0_.EMPLOYEE_NAME = 'b'))
```

**Reply by ironclad-15256304** [reply]
承知しました。ご回答頂き有難うございます。

**Reply by nakamura-to** [reply]
あ、私が提示した例が間違っていますね。ANDとORを入れ替えたものが正しいと思います。読み替えていただければと思います。

**Reply by ironclad-15256304** [reply]
有難うございます！

ちなみに後学のためにお伺いしたいのですが、Tuple2～9のうちTuple2のみしかin句でサポートされていないのはなにか理由があるのでしょうか？
おそらく今後もサポートされることはないという認識で問題ないのでしょうか。

**Reply by nakamura-to** [reply]
よく利用されると思われるケースのみをサポートしていました。
3列については #970 でサポートします。
追加のサポートについてはユーザーの皆さんの要望次第で検討したいと思います。

**Reply by ironclad-15256304** [reply]
ご回答およびご検討頂き有難うございます。3列に対応いただけるだけでもかなり実装の助けとなります！
お手数おかけしますが、よろしくお願い致します。

**Reply by nakamura-to** [reply]
リリースしました
https://github.com/domaframework/doma/releases/tag/2.54.0

**Reply by ironclad-15256304** [reply]
迅速なご対応有難うございます！！！
